src/test/scala/com/fulcrumgenomics/fastq/DemuxFastqsTest.scala

```diff
@@ -0,0 +1,524 @@
+    val samples: Seq[Sample] = {
+      val sampleSheet = SampleSheet(this.sampleSheetPath).map(s => withCustomSampleBarcode(s, "Sample_Barcode"))
+      sampleSheet.toSeq :+ DemuxFastqs.unmatchedSample(sampleSheet.size, structures)
+    }
+    samples.map { sample =>
+      val hdr = new SAMFileHeader()
+      hdr.addReadGroup(new SAMReadGroupRecord(sample.sampleId))
+      new SampleInfo(sample=sample, writer=null, isUnmatched=sample.sampleName==UnmatchedSampleId) {
+        override val header: SAMFileHeader = hdr
+      }
+    }
+  }
+
+  /** Helper method to create a [[FastqDemultiplexer]] */
+  private def dx(structures: Seq[ReadStructure], mm: Int = 2, md: Int = 1, mn: Int = 1): FastqDemultiplexer = {
+    new FastqDemultiplexer(sampleInfos=toSampleInfos(structures), readStructures=structures,
+      maxMismatches=mm, minMismatchDelta=md, maxNoCalls=mn)
+  }
+
+  private def fq(name: String, bases: String, readNumber: Option[Int]=None): FastqRecord = FastqRecord(name=name, bases=bases, quals="I"*bases.length, comment=None, readNumber=readNumber)
+
+  private def verifyFragUnmatchedSample(demuxRecord: DemuxRecord): Unit = {
+    demuxRecord.records.length shouldBe 1
+    val record = demuxRecord.records.headOption.value
+
+    record.getReadName shouldBe "frag"
+    record.getReadString shouldBe "A"*100
+    record.getBaseQualityString shouldBe "I"*100
+    record.getReadPairedFlag shouldBe false
+    record.getReadUnmappedFlag shouldBe true
+    record.getReadGroup.getId shouldBe UnmatchedSampleId
+    record.getAttribute("RX") shouldBe null
+  }
+
+  "FastqDemultiplexer" should "demux fragment reads" in {
+    val demuxer     = dx(structures=Seq(ReadStructure("17B100T")))
+    val fastqRecord = fq(name="frag", bases="AAAAAAAAGATTACAGA" + "A"*100)
+
+    val demuxRecord = demuxer.demultiplex(fastqRecord)
+    demuxRecord.records.length shouldBe 1
+    val record = demuxRecord.records.headOption.value
+
+    record.getReadName shouldBe "frag"
+    record.getReadString shouldBe "A"*100
+    record.getBaseQualityString shouldBe "I"*100
+    record.getReadPairedFlag shouldBe false
+    record.getReadUnmappedFlag shouldBe true
+    record.getReadGroup.getId shouldBe "20000101-EXPID-1"
+    record.getAttribute("RX") shouldBe null
+  }
+
+  it should "demux paired reads with in-line sample barcodes" in {
+    val demuxer = dx(structures=Seq(ReadStructure("8B100T"),ReadStructure("9B100T")))
+    val fq1 = fq(name="pair", bases="AAAAAAAA" + "A"*100)
+    val fq2 = fq(name="pair", bases="GATTACAGA" + "T"*100)
+
+    val demuxRecord = demuxer.demultiplex(fq1, fq2)
+    demuxRecord.records.length shouldBe 2
+    val r1 = demuxRecord.records.headOption.value
+    val r2 = demuxRecord.records.lastOption.value
+
+    r1.getReadName shouldBe "pair"
+    r1.getReadString shouldBe "A"*100
+    r1.getBaseQualityString shouldBe "I"*100
+    r1.getReadPairedFlag shouldBe true
+    r1.getFirstOfPairFlag shouldBe true
+    r1.getReadUnmappedFlag shouldBe true
```

Comment:
to be on the safe side you could also assert that `getSecondOfPairFlag` is `false`.